pkg/apis/crd/v1alpha1/types.go

```diff
@@ -291,3 +291,95 @@ type TLSProtocol struct {
 	// SNI (Server Name Indication) indicates the server domain name in the TLS/SSL hello message.
 	SNI string `json:"sni,omitempty"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// BGPPolicy defines BGP configuration applied to Nodes.
+type BGPPolicy struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard metadata of the object.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec BGPPolicySpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type BGPPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []BGPPolicy `json:"items"`
+}
+
+// BGPPolicySpec defines the specification for a BGP policy.
+type BGPPolicySpec struct {
+	// NodeSelector is utilized to select Nodes to which the policy is applied. It's important to note that only one
+	// BGPPolicy will be effective and enforced per Node, and others serve as alternatives.
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector"`
+
+	// LocalASN is the AS number used by the BGP protocol.
+	LocalASN int32 `json:"localASN"`
+
+	// ListenPort is the port on which the BGP protocol listens.
+	ListenPort *int32 `json:"listenPort,omitempty"`
+
+	// Advertisements selects IPs or CIDRs to be advertised to external BGP peers.
+	Advertisements Advertisements `json:"advertisements"`
+
+	// BGPPeers are the list of external BGP peers.
+	BGPPeers []BGPPeer `json:"bgpPeers"`
+}
+
+type Advertisements struct {
+	Service *ServiceAdvertisement `json:"service,omitempty"`
+
+	Pod *PodAdvertisement `json:"pod,omitempty"`
+
+	Egress *EgressAdvertisement `json:"egress,omitempty"`
+}
+
+type ServiceAdvertisement struct {
+	// Determine whether to advertise ClusterIPs.
+	ClusterIPs bool `json:"clusterIPs,omitempty"`
+
+	// Determine whether to advertise ExternalIPs.
+	ExternalIPs bool `json:"externalIPs,omitempty"`
+
+	// Determine whether to advertise LoadBalancerIPs.
+	LoadBalancerIPs bool `json:"loadBalancerIPs,omitempty"`
```

Comment:
I would personally make these pointers (`*bool`), similar to how the `Immutable` field is a `*bool` for Secrets and ConfigMaps in the K8s API. But I understand that this can be up for debate.

Reply:
My understanding is there are a few cases pointers are typically used:
1. Unset and zero value have different meanings
2. The field is not applicable and means nothing in some cases (really optional)
3. The field is added after the API is released, existing objects don't have it set. (but it seems not always the case)

In the booleans' cases, is there any reason why pointers are better? It seems to me:
- From usage's perspective, they are same
- From semantic perspective, they are applicable in the context of `ServiceAdvertisement`
- From code's perspective, we need to handle one more case when they are pointers

Reply:
I was going with this guideline from the K8s API conventions:

> Therefore, we ask that pointers always be used with optional fields that do not have a built-in nil value.

I see a few examples for core types: `Immutable` field in `ConfigMap`, `RunAsNonRoot` field in `PodSecurityContext`, ... even though in both of these cases the default value is `false`.

I agree that booleans are a very specific case, because there are only 2 possible values. The way I see it is that an optional boolean field should not need to be treated differently regardless of the case: 1) default value is `false`, 2) default value is `true` (only possible with pointer), 3) default value has its own meaning (only possible with pointer).

But since the use of booleans is somewhat discouraged in K8s API, did we consider having a slice of IP address types instead?

Reply:
> did we consider having a slice of IP address types instead?

Do you mean that the specific Service IPs (including clusterIPs, loadBalancerIPs and externalIPs) to be advertised are stored in a slice, instead of multiple booleans to determine whether to advertise the clusterIPs, loadBalancerIPs or externalIPs of all Services?

Reply:
> I was going with this guideline from the K8s API conventions:
> 
> > Therefore, we ask that pointers always be used with optional fields that do not have a built-in nil value.
> 
> I see a few examples for core types: `Immutable` field in `ConfigMap`, `RunAsNonRoot` field in `PodSecurityContext`, ... even though in both of these cases the default value is `false`.

I think many such fields are pointers because they were [new fields added to existing APIs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version) and gated by a feature gate, e.g. [`Immutable` is added along with ImmutableEphemeralVolumes](https://github.com/kubernetes/kubernetes/commit/e612ebfdff22e4bd27ad8345f7c82f074bfedf26). Using pointers ensures:
* The field is entirely absent from API responses when empty
* When the feature is disabled, user input can be cleared.
* When the feature is enabled, only user input is persisted in the storage.

[`RunAsNonRoot` was changed to pointers](https://github.com/kubernetes/kubernetes/commit/393e2bc0195343c83933aa0cf1c02e854fe8779b#diff-41f03de739aac2ef36071b5a8ceefd4d8400b34f8eec6420941ab99e8ef48185) because it really became optional and can be set in two places, one takes precedence over the other, so it needs to know whether users input it.

In our case, I think these booleans are not really "optional", we need to know if these IPs should be advertised, we just don't need users to input them when they don't need to advertise them.

Reply:
We can remove the pointers. But any thought on having a list of IP types?
```go
type ServiceIPType string
const (
        ServiceIPTypeClusterIP ServiceIPType = "ClusterIP"
        ServiceIPTypeLoadBalancerIP ServiceIPType = "LoadBalancerIP"
        ServiceIPTypeExternalIP ServiceIPType = "ExternalIP"
)

type ServiceAdvertisement struct {
        IPTypes []ServiceIPType `json:"ipTypes,omitempty"`
}
```

I am not sure it is less convenient for users actually:
```yaml
ipTypes: [LoadBalancerIP, ExternalIP]
```

Reply:
The list of types looks good to me. I'm not sure which is better either.

Reply:
I'm fine with the list, and it seems that it is not very inconvenient to use it.